### PR TITLE
Feature/search microservice

### LINKED ISSUE #1737: Create a search microservice
We are moving toward microservices for our back-end, and I want to refactor our current search routes into a separate service.

We have some existing code for this at https://github.com/Seneca-CDOT/telescope/blob/master/src/backend/web/routes/query.js and a working implementation for searching Posts and Authors.

To begin, let's duplicate what we have, but provide a better REST API, maybe something like this:

- `GET /search/user/?q=...&page=..&per_page=..` - allows searching for user's by the given `q` (i.e., query) string.  This probably should search their names (first, last, display, github username).  It should return paged results (e.g., `page` results with `per_page` results per page).  See the [GitHub Search API](https://docs.github.com/en/rest/reference/search#search-commits) for examples of what this might look like.

- `GET /search/post/?q=...&page=..&per_page=..` - similar idea to the above, but let's you query with post data (title, body of post) and returns results.  We might want to expand this API a bit to support things that Elastic can do, since it is a lot more powerful with it's query API.  That could happen later, though.

There are likely other things we can add search endpoints for, but would be a good start.

### PR DESCRIPTION
**Pull request by izhuravlev**
## Issue This PR Addresses
Updated #1802 
Closes #1737 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [X] **New Feature**: Change which adds functionality
- [X] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

Adds support for `search.docker.localhost`, which ports the original ELK query controller to a dedicated service. The following is added as a warning to not work further on the original codebase for querying: 

NOTE: The tests will be done in the follow-up to this PR. 

```javascript
/*
 * NOTE: This has been ported to ~/src/api/search
 * Future updates will remove the codebase below
 * and point towards `service.docker.localhost` for querying.
 */
 ```

image

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: config/env.development [hunk 1]
```diff
@@ -89,6 +89,15 @@ IMAGE_PORT=4444
...
 
+################################################################################
+# Search Service
+################################################################################
+
+# Search Service Port (default is 4445)
+SEARCH_PORT=4445
+
+# Search Service URL
+SEARCH_URL=http://api.telescope.localhost/v1/search
```
Comment (humphd) [inline config/env.development:100]: `http://localhost/v1/search`

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -69,6 +69,34 @@ services:
...
       - 'traefik.http.routers.auth.middlewares=strip_auth_prefix'
 
+  # search service
+  search:
+    container_name: 'search'
+    build:
+      context: ../src/api/search
+      dockerfile: Dockerfile
+    environment:
+      - ELASTIC_APM_SERVICE_NAME=search
```
Comment (humphd) [inline docker/docker-compose.yml:79]: We don't have APM yet, so remove this.

File: docker/docker-compose.yml [hunk 3]
```diff
@@ -69,6 +69,34 @@ services:
...
+      - SEARCH_PORT
+      - ELASTICSEARCH_HOSTS=http://elasticsearch:9200
```
Comment (humphd) [inline docker/docker-compose.yml:81]: I think this is not needed now, we use ELASTIC_URL and _PORT.  Remove.

File: docker/docker-compose.yml [hunk 4]
```diff
@@ -69,6 +69,34 @@ services:
...
+      - ELASTIC_APM_SERVER_URL=http://apm:8200
```
Comment (humphd) [inline docker/docker-compose.yml:82]: This can go (no APM yet).

File: src/api/search/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,19 @@
+FROM node:lts-alpine as base
+
+# https://snyk.io/blog/10-best-practices-to-containerize-nodejs-web-applications-with-docker/
+RUN apk add dumb-init
+
+# TODO: Add feeding in the port from a .env file
```
Comment (humphd) [inline src/api/search/Dockerfile:6]: Remove this comment.

File: src/api/search/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+WORKDIR /app
+
+COPY --chown=node:node . .
+
+RUN npm install --only=production --no-package-lock
+
+USER node
+
+CMD ["dumb-init", "node", "server.js"]
```
Comment (humphd) [inline src/api/search/Dockerfile:19]: "src/server.js"

File: src/api/search/package.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@senecacdot/search-service",
+  "private": true,
+  "version": "1.0.0",
+  "description": "A service for optimizing images",
+  "scripts": {
+    "dev": "nodemon src/server.js",
+    "start": "node server.js"
```
Comment (chrispinkney) [inline src/api/search/package.json:8]:
```suggestion
    "start": "node src/server.js"
```

File: src/api/search/package.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    "@elastic/elasticsearch-mock": "^0.3.0",
+    "@senecacdot/satellite": "^1.x",
+    "express-validator": "^6.9.2"
+  },
+  "engines": {
+    "node": ">=12.0.0"
+  },
+  "devDependencies": {
+    "nodemon": "^2.0.7",
+    "supertest": "^6.1.3"
```
Comment (humphd) [inline src/api/search/package.json:27]: Remove this for now.

**Review by chrispinkney — CHANGES_REQUESTED** [review]

File: docker/docker-compose.yml [hunk 9]
```diff
@@ -69,6 +69,34 @@ services:
       - 'traefik.http.middlewares.strip_auth_prefix.stripprefix.forceSlash=true'
       - 'traefik.http.routers.auth.middlewares=strip_auth_prefix'
 
+  # search service
+  search:
+    container_name: 'search'
+    build:
+      context: ../src/api/search
+      dockerfile: Dockerfile
+    environment:
```
Comment (humphd) [inline docker/docker-compose.yml:78]: Inside this environemnt, you need to make sure all the `process.env` values you use are defined here, for example:

- ELASTIC_MAX_RESULTS_PER_PAGE
- ELASTIC_URL
- ELASTIC_PORT

I'm not sure if there are more?

Comment (humphd) [reply docker/docker-compose.yml:78]: In `production.yml` you would override this:


      - ELASTIC_PORT=9200
      - ELASTIC_URL=http://elasticsearch

Comment (humphd) [reply docker/docker-compose.yml:78]: In `development.yml` you would do:

```
ELASTIC_URL=http://localhost
```

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/search/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+## Install
+
+```
+npm install
+```
+
+## Usage
+
+```
+# normal mode
```
Comment (chrispinkney) [inline src/api/search/README.md:14]: Do you want to add a section about [running in dev mode](https://github.com/Seneca-CDOT/telescope/blob/master/src/api/users/README.md#usage)?

File: src/api/search/README.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
+# Search Service
+
+The Search Service provides a direct querying controller which interfaces with our ELK stack.
+
+## Install
+
+```
+npm install
```
Comment (chrispinkney) [inline src/api/search/README.md:8]: or `install:search-service`?

**Review by chrispinkney — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
This needs a rebase to deal with the Parser service landing just now.

File: src/api/search/README.md (hunk 11 above)
Comment (chrispinkney) [reply src/api/search/README.md:8]: If we're trying to rush this in now I'm okay with fixing this after.

File: src/api/search/README.md (hunk 10 above)
Comment (izhuravlev) [reply src/api/search/README.md:14]: I don't have anything on this cos I can't even run it myself yet...

File: src/api/search/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,17 @@
+FROM node:lts-alpine as base
+
+# https://snyk.io/blog/10-best-practices-to-containerize-nodejs-web-applications-with-docker/
+RUN apk add dumb-init
+
+# Force production env, regardless of what we get from .env
+ENV NODE_ENV production
```
Comment (humphd) [inline src/api/search/Dockerfile:7]: You can remove it.

**Review by humphd — APPROVED** [review]

**Review by humphd — APPROVED** [review]

**Review by PedroFonsecaDEV — APPROVED** [review]

**Review by chrispinkney — APPROVED** [review]

**Pull request merged**